backbone/src/test/java/org/researchstack/backbone/ui/task/TaskViewModelTest.kt

```diff
@@ -0,0 +1,419 @@
+package org.researchstack.backbone.ui.task
+
+import android.app.Application
+import android.content.Intent
+import com.nhaarman.mockitokotlin2.*
+
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mockito
+import org.researchstack.backbone.step.Step
+import org.researchstack.backbone.task.Task
+import org.junit.rules.TestRule
+import org.junit.Rule
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import junit.framework.TestCase.assertEquals
+import org.junit.Assert
+import org.junit.runner.RunWith
+import org.mockito.Mock
+import org.mockito.Mockito.`when`
+import org.mockito.junit.MockitoJUnitRunner
+import org.researchstack.backbone.result.StepResult
+import org.researchstack.backbone.result.TaskResult
+
+@RunWith(MockitoJUnitRunner::class)
+class TaskViewModelTest {
+
+    @Rule
+    @JvmField
+    val rule: TestRule = InstantTaskExecutorRule()
+
+    @Mock
+    private var application: Application? = null
+    @Mock
+    private var intent: Intent? = null
+    @Mock
+    private var task: Task? = null
+
+    private lateinit var taskViewModel: TaskViewModel
+
+    @Before
+    fun before() {
+        whenever(intent?.getSerializableExtra(TaskActivity.EXTRA_TASK)).thenReturn(eq(task))
+    }
+
+    @Test
+    fun testInitViewModel_TaskShouldNotBeEmpty_EditModeMustBeFalse() {
+        // Assemble
+        // Act
+        taskViewModel = createViewModel()
+
+        // Assert
+        assertEquals(taskViewModel.editing, false)
+        assertEquals(taskViewModel.task, task)
+    }
+
+    @Test
+    fun testEdit_editShouldBeTrue_EditStepShouldBeUpdated_CancelMenuShouldBeTrueToHideCancelInTitleBar() {
+        // Assemble
+        val selectedEditStep = createCurrentStep("selectedEditStep")
+        taskViewModel = createViewModel()
+
+        // Act
+        taskViewModel.edit(selectedEditStep)
+
+        // Assert
+        Assert.assertEquals(taskViewModel.currentStep, selectedEditStep)
+        Assert.assertEquals(taskViewModel.editStep.value, selectedEditStep)
+        Assert.assertEquals(taskViewModel.editing, true)
+        Assert.assertEquals(taskViewModel.hideMenuItemCancel.value, true)
+    }
+
+
+    @Test
+    fun testRemoveUpdatedLayout_editShouldBeFalse_EditStepShouldBeUpdated_CancelMenuShouldBeFalseToShowCancelInTitleBar() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
```

Comment:
Why do you need to `spy` the subject here?

Reply:
Here I want to override isReviewStep fun, and test  updateCancelEditInLayout + hideMenuItemCancel live date